core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +118,135 @@ where
         }
     }
 
     /// Returns the root hash and the Merkle proof for a leaf with the specified 0-based `index`.
-    pub fn merkle_root_and_path(self, index: usize) -> (H256, Vec<H256>) {
-        let mut merkle_path = vec![];
-        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut merkle_path));
-        (root_hash, merkle_path)
+    /// `index` is relative to the leftmost uncached leaf.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        let mut end_path = vec![];
+        let root_hash = self.compute_merkle_root_and_path(index, None, Some(&mut end_path));
+        (root_hash, end_path)
+    }
+
+    /// Returns the root hash and the Merkle proofs for a range of leafs.
+    /// The range is 0..length, where `0` is the leftmost untrimmed leaf.
+    pub fn merkle_root_and_paths_for_range(&self, length: usize) -> (H256, Vec<H256>, Vec<H256>) {
+        let (mut left_path, mut right_path) = (vec![], vec![]);
+        let root_hash = self.compute_merkle_root_and_path(
+            length - 1,
+            Some(&mut left_path),
+            Some(&mut right_path),
+        );
+        (root_hash, left_path, right_path)
+    }
+
+    /// Adds a raw hash to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push_hash(&mut self, leaf_hash: H256) {
+        self.hashes.push_back(leaf_hash);
+        if self.start_index + self.hashes.len() > self.binary_tree_size {
+            self.binary_tree_size *= 2;
+        }
+    }
+
+    /// Adds a new leaf to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push(&mut self, leaf: [u8; LEAF_SIZE]) {
+        let leaf_hash = self.hasher.hash_bytes(&leaf);
+        self.push_hash(leaf_hash);
+    }
+
+    /// Caches the rightmost `count` leaves.
+    /// Does not affect the root hash, but makes it impossible to get the paths to the cached leaves.
+    /// # Panics
+    /// Panics if `count` is greater than the number of non-cached leaves in the tree.
+    pub fn trim_start(&mut self, count: usize) {
+        assert!(self.hashes.len() >= count, "not enough leaves to cache");
+        let mut new_cache = vec![];
+        // Cache is a subset of the path to the first untrimmed leaf.
+        let root = self.compute_merkle_root_and_path(count, None, Some(&mut new_cache));
+        self.hashes.drain(..count);
+        self.start_index += count;
+        // It is important to add the root in case we just trimmed all leaves *and*
+        // the tree will grow on the next push.
+        new_cache.push(root);
+        self.cache = new_cache;
     }
 
     fn compute_merkle_root_and_path(
-        self,
-        mut index: usize,
-        mut merkle_path: Option<&mut Vec<H256>>,
+        &self,
+        mut end_index: usize,
+        mut start_path: Option<&mut Vec<H256>>,
```

Comment:
Dumb question: Since the relative start index is always zero, doesn't the `start_path` always *conceptually* equal to `cache` (except for the root hash at the end of `cache`)? See also the comment about partial Merkle paths; the paths are not exactly equal right now due to the differing right subset of the path (which can change when new elements are inserted to the tree after `cache` is computed), but the right subset should not be used in proof verification anyway.

Reply:
Good catch